### ISSUE TITLE
Formalize shim <-> runtime initial protocol

### ISSUE THREAD

**Issue by sameo** [01org/cc-oci-runtime#559]
When exchanging the shim arguments before `execve`'ing the shim binary, we use a generic args pipe where we push the proxy IO fd and base sequence number in a predefined, undocumented order. We also have a separate pipe for pushing potential errors.

We may want to have a slightly more structured protocol here that would allow us to not depend on any specific order and get rid of the error pipe:

- One single pipe for both data and error
- A more structured data structure for passing proxy IO fd, base and errors:
```
struct shim_args {
          unsigned int type;
          int data;
};
```

where `type` could be:

```
#define SHIM_ARGS_IO_FD 0
#define SHIM_ARGS_IO_BASE 1
#define SHIM_ARGS_ERROR 2
```

This would simplify the code and would allow us to actually document that protocol for those who'd like to use cc-shim as an independent process.

**Comment by amshinde** [comment]
@sameo Is this meant for code simplification for 2.1?
The order of passing the arguments to the forked process is totally specific to the runtime. Another process(runtime) could choose to do this differently. The shim is ultimately exec-ed with the required arguments. So I am not very clear about how the above would impact this : "would allow the shim to run as an independent process"

**Comment by devimc** [comment]
Will be fixed in 3.0 - https://github.com/clearcontainers/runtime
